File: account_nuker.py
import aiohttp
import discord
import asyncio
import logging
from datetime import datetime
from tqdm import tqdm

logging.basicConfig(
    filename=f'account_nuker.log',
    filemode='w',
    format='%(asctime)s - %(levelname)s - %(message)s',
    level=logging.INFO
)

intents = discord.Intents.all()
client = discord.Client(intents=intents, self_bot=True)
token = None

def display_banner():
    print("""                                                                                                                
    █████████                                                     █████       ██████   █████            █████                        
  ███░░░░░███                                                   ░░███       ░░██████ ░░███            ░░███                         
 ░███    ░███   ██████   ██████   ██████  █████ ████ ████████   ███████      ░███░███ ░███  █████ ████ ░███ █████  ██████  ████████ 
 ░███████████  ███░░███ ███░░███ ███░░███░░███ ░███ ░░███░░███ ░░░███░       ░███░░███░███ ░░███ ░███  ░███░░███  ███░░███░░███░░███
 ░███░░░░░███ ░███ ░░░ ░███ ░░░ ░███ ░███ ░███ ░███  ░███ ░███   ░███        ░███ ░░██████  ░███ ░███  ░██████░  ░███████  ░███ ░░░ 
 ░███    ░███ ░███  ███░███  ███░███ ░███ ░███ ░███  ░███ ░███   ░███ ███    ░███  ░░█████  ░███ ░███  ░███░░███ ░███░░░   ░███     
 █████   █████░░██████ ░░██████ ░░██████  ░░████████ ████ █████  ░░█████     █████  ░░█████ ░░████████ ████ █████░░██████  █████    
░░░░░   ░░░░░  ░░░░░░   ░░░░░░   ░░░░░░    ░░░░░░░░ ░░░░ ░░░░░    ░░░░░     ░░░░░    ░░░░░   ░░░░░░░░ ░░░░ ░░░░░  ░░░░░░  ░░░░░     
                                                      by ThunderDoesDev
""")
    print("---------------------------------------------------")
    print("                 OPTIONS MENU                     ")
    print("---------------------------------------------------")
    print("1. Show Account Info")
    print("2. Leave or Delete Guilds")
    print("3. Delete Friends")
    print("4. Delete DMs")
    print("5. Full Nuke (All actions)")
    print("---------------------------------------------------")

async def display_info():
    logging.info("Displaying account info.")
    print(f"\nAccount Information:")
    print(f"Username: {client.user.name}#{client.user.discriminator}")
    print(f"User ID: {client.user.id}")
    print(f"Email: {client.user.email if client.user.email else 'No email set'}")
    print(f"Phone: {client.user.phone if client.user.phone else 'No phone number set'}")
    print(f"Verified: {client.user.verified}")
    logging.info(f"Account Info: {client.user.name}#{client.user.discriminator}, ID: {client.user.id}")
    print("\nPayment Sources:")
    try:
        payment_sources = await client.http.request(discord.http.Route('GET', '/users/@me/billing/payment-sources'))
        if payment_sources:
            for source in payment_sources:
                print(f"Payment Type: {source['type']}")
                print(f"Last 4 Digits: {source['last_4']}")
                print(f"Expires: {source['expires_month']}/{source['expires_year']}")
                print(f"Billing Address: {source['billing_address']['country']}")
                logging.info(f"Payment Source: {source['type']}, Last 4: {source['last_4']}, Expires: {source['expires_month']}/{source['expires_year']}, Billing Country: {source['billing_address']['country']}")
        else:
            print("No payment sources found.")
            logging.info("No payment sources found.")
    except Exception as e:
        logging.error(f"Failed to retrieve payment sources: {str(e)}")
    print("\nBilling Information:")
    try:
        billing_info = await client.http.request(discord.http.Route('GET', '/users/@me/billing/payment-sources'))
        if billing_info:
            for info in billing_info:
                print(f"Billing Type: {info['type']}")
                print(f"Country: {info['billing_address']['country']}")
                print(f"Last 4 Digits: {info['last_4']}")
                print(f"Expires: {info['expires_month']}/{info['expires_year']}")
                logging.info(f"Billing Info: Type: {info['type']}, Country: {info['billing_address']['country']}, Last 4: {info['last_4']}, Expires: {info['expires_month']}/{info['expires_year']}")
        else:
            print("No billing information found.")
            logging.info("No billing information found.")
    except Exception as e:
        logging.error(f"Failed to retrieve billing information: {str(e)}")
    print("\nSubscription Information:")
    try:
        subscriptions = await client.http.request(discord.http.Route('GET', '/users/@me/billing/subscriptions'))
        if subscriptions:
            for sub in subscriptions:
                print(f"Subscription Type: {sub['type']}")
                print(f"Status: {sub['status']}")
                print(f"Current Period End: {sub['current_period_end']}")
                logging.info(f"Subscription: Type: {sub['type']}, Status: {sub['status']}, Period End: {sub['current_period_end']}")
        else:
            print("No active subscriptions.")
            logging.info("No active subscriptions.")
    except Exception as e:
        logging.error(f"Failed to retrieve subscription information: {str(e)}")
    print("\nReturning to the main menu...\n")
    main()

async def fetch_guilds(token):
    base_url = "https://discord.com/api/v9"
    headers = {"Authorization": token}
    async with aiohttp.ClientSession() as session:
        response = await session.get(f"{base_url}/users/@me/guilds", headers=headers)
        if response.status == 200:
            guilds = await response.json()
            if guilds:
                logging.info(f"Fetched {len(guilds)} guilds.")
                return guilds
            else:
                logging.warning("No guilds found.")
                return None
        else:
            logging.error(f"Failed to fetch guilds (status code: {response.status})")
            return None

async def leave_guild(token, guild):
    guild_id = str(guild['id'])
    guild_name = guild['name']
    headers = {"Authorization": token}
    base_url = f"https://discord.com/api/v9"    
    async with aiohttp.ClientSession() as session:
        response = await session.delete(f"{base_url}/users/@me/guilds/{guild_id}", headers=headers)    
    if response.status in [200, 204]:
        logging.info(f"Left guild: {guild_name}")
        return True
    elif response.status == 429:
        retry_after = await response.json()
        logging.warning(f"Rate limited. Retrying after {retry_after['retry_after']} seconds...")
        await asyncio.sleep(retry_after['retry_after'])
        return await leave_guild(token, guild)  # Proper recursive call with token and guild
    else:
        error_text = await response.text()
        logging.error(f"Failed to leave guild: {guild_name} (status code: {response.status}, error: {error_text})")
        return False

async def delete_guild(session, guild, headers):
    guild_id = guild['id']
    guild_name = guild['name']
    url = f"https://discord.com/api/v9/guilds/{guild_id}"
    response = await session.delete(url, headers=headers)
    if response.status in [200, 204]:
        logging.info(f"Deleted guild: {guild_name}")
        return True
    elif response.status == 429:
        retry_after = await response.json()
        logging.warning(f"Rate limited. Retrying after {retry_after['retry_after']} seconds...")
        await asyncio.sleep(retry_after['retry_after'])
        return await delete_guild(session, guild, headers)
    else:
        logging.error(f"Failed to delete guild: {guild_name} (status code: {response.status})")
        return False

async def process_guilds(token):
    guilds = await fetch_guilds(token)
    if guilds and len(guilds) > 0:
        leave_count, delete_count = 0, 0        
        async with aiohttp.ClientSession() as session:
            headers = {"Authorization": token}
            leave_guilds = [guild for guild in guilds if not guild['owner']]
            leave_results = []
            for guild in tqdm(leave_guilds, desc="Leaving guilds"):
                result = await leave_guild(token, guild)
                leave_results.append(result)
                leave_count += result
            logging.info(f"Processed {len(leave_results)} leave actions, Success: {leave_count}")
            delete_guilds = [guild for guild in guilds if guild['owner']]
            delete_results = []
            for guild in tqdm(delete_guilds, desc="Deleting guilds"):
                result = await delete_guild(session, guild, headers)
                delete_results.append(result)
                delete_count += result
            logging.info(f"Processed {len(delete_results)} delete actions, Success: {delete_count}")
    else:
        logging.info("No guilds to process.")

async def close_all_dms(token):
    headers = {"Authorization": token, "Content-Type": "application/json"}
    base_url = "https://discord.com/api/v9"
    async with aiohttp.ClientSession() as session:
        try:
            async with session.get(f"{base_url}/users/@me/channels", headers=headers) as response:
                if response.status == 200:
                    dms_data = await response.json()
                    total_dms = len(dms_data)
                    deleted_dms = 0
                    print(f"Deleting {total_dms} DMs...")
                    for dm in tqdm(dms_data, desc="Deleting DMs"):
                        try:
                            channel_id = dm['id']
                            async with session.delete(f"{base_url}/channels/{channel_id}", headers=headers) as delete_response:
                                if delete_response.status in [200, 204]:
                                    logging.info(f"Deleted DM: {channel_id}")
                                    deleted_dms += 1
                                elif delete_response.status == 420:  
                                    retry_after = await delete_response.json()  
                                    logging.warning(f"Rate limited. Retrying after {retry_after['retry_after']} seconds...")
                                    await asyncio.sleep(retry_after['retry_after'])
                                    await close_all_dms(session, base_url, headers, token) 
                                else:
                                    logging.error(f"Failed to delete DM: {channel_id} (status: {delete_response.status})")
                        except Exception as e:
                            logging.error(f"Error deleting DM {channel_id}: {str(e)}")
                    if deleted_dms == total_dms:
                        logging.info("All DMs were successfully deleted.")
                    else:
                        logging.warning(f"Only {deleted_dms} out of {total_dms} DMs were deleted.")
                else:
                    logging.error(f"Failed to fetch DMs (status code: {response.status})")
        except Exception as e:
            logging.error(f"Error fetching DMs: {str(e)}")

async def remove_all_friends(token):
    headers = {"Authorization": token, "Content-Type": "application/json"}
    base_url = "https://discord.com/api/v9"
    async with aiohttp.ClientSession() as session:
        try:
            async with session.get(f"{base_url}/users/@me/relationships", headers=headers) as response:
                if response.status == 200:
                    friends_data = await response.json()
                    total_friends = len(friends_data)
                    removed_friends = 0
                    print(f"Removing {total_friends} friends...")
                    for friend in tqdm(friends_data, desc="Removing Friends"):
                        try:  
                            if friend['type'] == 1:
                                friend_id = friend['id']
                                async with session.delete(f"{base_url}/users/@me/relationships/{friend_id}", headers=headers) as delete_response:
                                    if delete_response.status in [200, 204]:
                                        removed_friends += 1
                                        logging.info(f"Removed friend: {friend['user']['username']}")
                                    elif delete_response.status == 420:  
                                        retry_after = await delete_response.json()  
                                        logging.warning(f"Rate limited. Retrying after {retry_after['retry_after']} seconds...")
                                        await asyncio.sleep(retry_after['retry_after'])
                                        await remove_all_friends(session, base_url, headers, token) 
                                    else:
                                        logging.error(f"Failed to remove friend: {friend['user']['username']} (status: {delete_response.status})")
                        except Exception as e:
                            logging.error(f"Error removing friend {friend['user']['username']}: {str(e)}")
                    if removed_friends == total_friends:
                        logging.info("All friends were successfully removed.")
                        print("All friends were successfully removed.")
                    else:
                        logging.warning(f"Only {removed_friends} out of {total_friends} friends were removed.")
                        print(f"Only {removed_friends} out of {total_friends} friends were removed.")
                else:
                    logging.error(f"Failed to fetch friends (status code: {response.status})")
        except Exception as e:
            logging.error(f"Error fetching friends: {str(e)}")

async def nuke_all():
    logging.info("Executing full nuke (servers, DMs, friends, profile).")
    print("Nuking everything...")
    await process_guilds(token)
    await close_all_dms(token)
    await remove_all_friends(token)

def confirm_choice():
    confirmation = input("Type 'Confirm' to process or 'Cancel' to return to the main menu: ")
    if confirmation.lower() == "confirm":
        return True
    elif confirmation.lower() == "cancel":
        print("Returning to the main menu...")
        logging.info("User canceled the operation.")
        main()
    else:
        print("Invalid input. Returning to main menu.")
        logging.warning("Invalid input during confirmation.")
        main()

async def nuke_options(option, token):
    try:
        if option == '1':
            await display_info()
        elif option == '2':
            await process_guilds(token)
        elif option == '3':
            await remove_all_friends(token)
        elif option == '4':
             await close_all_dms(token)
        elif option == '5':
            await nuke_all()
    except Exception as e:
        logging.error(f"An unexpected error occurred: {str(e)}")
    print("\nReturning to the main menu...\n")
    main()

@client.event
async def on_ready():
    logging.info(f'Logged in as {client.user}')
    print(f'Logged in as {client.user}')
    await nuke_options(option, token)

# Main flow
def main():
    display_banner()
    global option
    option = input("Choose an option (1-5): ")
    global token
    token = input("Enter your token: ")
    if confirm_choice():
        try:
            client.run(token, bot=False)
        except Exception as e:
            logging.error(f"Failed to log in: {str(e)}")

if __name__ == "__main__":
    main()

File: account_nuker.js
const axios = require('axios');
const readline = require('readline');
const fs = require('fs');
const cliProgress = require('cli-progress');

// Logging setup
const logStream = fs.createWriteStream('account_nuker.log', { flags: 'w' });
const log = (message) => {
    const timestamp = new Date().toISOString();
    logStream.write(`[${timestamp}] ${message}\n`);
    console.log(message);
};

let token = null;
const baseURL = 'https://discord.com/api/v9';

// Banner display
function displayBanner() {
    console.log(`
█████████                                                     █████       ██████   █████            █████                        
███░░░░░███                                                   ░░███       ░░██████ ░░███            ░░███                         
███    ░███   ██████   ██████   ██████  █████ ████ ████████   ███████      ░███░███ ░███  █████ ████ ░███ █████  ██████  ████████ 
███████████  ███░░███ ███░░███ ███░░███░░███ ░███ ░░███░░███ ░░░███░       ░███░░███░███ ░░███ ░███  ░███░░███  ███░░███░░███░░███
███░░░░░███ ░███ ░░░ ░███ ░░░ ░███ ░███ ░███ ░███  ░███ ░███   ░███        ░███ ░░██████  ░███ ░███  ░██████░  ░███████  ░███ ░░░ 
███    ░███ ░███  ███░███  ███░███ ░███ ░███ ░███  ░███ ░███   ░███ ███    ░███  ░░█████  ░███ ░███  ░███░░███ ░███░░░   ░███     
█████   █████░██████ ░░██████ ░░██████  ░░████████ ████ █████  ░░█████     █████  ░░█████ ░░████████ ████ █████░░██████  █████    
░░░░░   ░░░░░  ░░░░░░   ░░░░░░   ░░░░░░    ░░░░░░░░ ░░░░ ░░░░░    ░░░░░     ░░░░░    ░░░░░   ░░░░░░░░ ░░░░ ░░░░░  ░░░░░░  ░░░░░     
                                                    by ThunderDoesDev
---------------------------------------------------
                 OPTIONS MENU
---------------------------------------------------
1. Show Account Info
2. Leave or Delete Guilds
3. Delete Friends
4. Delete DMs
5. Full Nuke (All actions)
---------------------------------------------------
`);
}

async function displayInfo() {
    log("Displaying account info.");
    let retry = true;
    while (retry) {
        try {
            const userResponse = await axios.get(`${baseURL}/users/@me`, {
                headers: { Authorization: token }
            });
            const user = userResponse.data;
            console.log(`\nAccount Information:`);
            console.log(`Username: ${user.username}#${user.discriminator}`);
            console.log(`User ID: ${user.id}`);
            console.log(`Email: ${user.email || 'No email set'}`);
            console.log(`Phone: ${user.phone || 'No phone number set'}`);
            console.log(`Verified: ${user.verified}`);
            log(`Account Info: ${user.username}#${user.discriminator}, ID: ${user.id}, Verified: ${user.verified}`);            
            const paymentResponse = await axios.get(`${baseURL}/users/@me/billing/payment-sources`, {
                headers: { Authorization: token }
            });
            const paymentSources = paymentResponse.data;
            if (paymentSources.length > 0) {
                console.log("\nPayment Sources:");
                paymentSources.forEach((source) => {
                    console.log(`Payment Type: ${source.type}`);
                    console.log(`Last 4 Digits: ${source.last_4}`);
                    console.log(`Expires: ${source.expires_month}/${source.expires_year}`);
                    console.log(`Billing Country: ${source.billing_address.country}`);
                    log(`Payment Source: ${source.type}, Last 4: ${source.last_4}`);
                });
            } else {
                console.log("No payment sources found.");
                log("No payment sources found.");
            }
            const subscriptionResponse = await axios.get(`${baseURL}/users/@me/billing/subscriptions`, {
                headers: { Authorization: token }
            });
            const subscriptions = subscriptionResponse.data;
            if (subscriptions.length > 0) {
                console.log("\nSubscription Information:");
                subscriptions.forEach((sub) => {
                    console.log(`Subscription Type: ${sub.type}`);
                    console.log(`Status: ${sub.status}`);
                    console.log(`Current Period End: ${new Date(sub.current_period_end).toLocaleDateString()}`);
                    log(`Subscription: Type: ${sub.type}, Status: ${sub.status}, Period End: ${sub.current_period_end}`);
                });
            } else {
                console.log("No subscriptions found.");
                log("No subscriptions found.");
            }
            retry = false;
        } catch (error) {
            if (error.response && error.response.status === 429) {
                const retryAfter = error.response.headers['retry-after'];
                log(`Rate limit hit. Retrying after ${retryAfter} seconds...`);
                await new Promise((resolve) => setTimeout(resolve, retryAfter * 1000));
            } else {
                log(`Error fetching account info: ${error.message}`);
                retry = false;
            }
        }
    }
}

async function processGuilds() {
    try {
        const response = await axios.get(`${baseURL}/users/@me/guilds`, {
            headers: { Authorization: token }
        });
        const guilds = response.data;
        let leaveCount = 0;
        let deleteCount = 0;
        const guildBar = new cliProgress.SingleBar({}, cliProgress.Presets.shades_classic);
        guildBar.start(guilds.length, 0);
        for (const guild of guilds) {
            let retry = true;
            while (retry) {
                try {
                    if (!guild.owner) {
                        // Leave guild
                        await axios.delete(`${baseURL}/users/@me/guilds/${guild.id}`, {
                            headers: { Authorization: token }
                        });
                        leaveCount++;
                        log(`Left guild: ${guild.name}`);
                    } else {
                        // Delete guild
                        await axios.delete(`${baseURL}/guilds/${guild.id}`, {
                            headers: { Authorization: token }
                        });
                        deleteCount++;
                        log(`Deleted guild: ${guild.name}`);
                    }
                    retry = false;
                } catch (error) {
                    if (error.response && error.response.status === 429) {
                        const retryAfter = error.response.headers['retry-after'];
                        log(`Rate limit hit. Retrying after ${retryAfter} seconds...`);
                        await new Promise((resolve) => setTimeout(resolve, retryAfter * 1000));
                    } else {
                        log(`Error processing guild: ${error.message}`);
                        retry = false;
                    }
                }
            }
            guildBar.increment();
        }
        guildBar.stop();
        log(`Processed ${leaveCount} leave actions and ${deleteCount} delete actions.`);
    } catch (error) {
        log(`Error processing guilds: ${error.message}`);
    }
}

async function confirmChoice() {
    return new Promise((resolve) => {
        rl.question("Type 'Confirm' to process or 'Cancel' to return to the main menu: ", (input) => {
            if (input.toLowerCase() === 'confirm') {
                resolve(true);
            } else if (input.toLowerCase() === 'cancel') {
                console.log("Returning to the main menu...");
                log("User canceled the operation.");
                resolve(false);
            } else {
                console.log("Invalid input. Returning to main menu.");
                log("Invalid input during confirmation.");
                resolve(false);
            }
        });
    });
}

async function closeAllDMs() {
    try {
        const response = await axios.get(`${baseURL}/users/@me/channels`, {
            headers: { Authorization: token }
        });
        const dms = response.data;
        const dmBar = new cliProgress.SingleBar({}, cliProgress.Presets.shades_classic);
        dmBar.start(dms.length, 0);
        for (const dm of dms) {
            let retry = true;
            while (retry) {
                try {
                    await axios.delete(`${baseURL}/channels/${dm.id}`, {
                        headers: { Authorization: token }
                    });
                    log(`Deleted DM with ID: ${dm.id}`);
                    retry = false;
                } catch (error) {
                    if (error.response && error.response.status === 429) {
                        const retryAfter = error.response.headers['retry-after'];
                        log(`Rate limit hit. Retrying after ${retryAfter} seconds...`);
                        await new Promise((resolve) => setTimeout(resolve, retryAfter * 1000));
                    } else {
                        log(`Error deleting DM: ${error.message}`);
                        retry = false;
                    }
                }
            }
            dmBar.increment();
        }
        dmBar.stop();
    } catch (error) {
        log(`Error closing DMs: ${error.message}`);
    }
}

async function removeAllFriends() {
    try {
        const response = await axios.get(`${baseURL}/users/@me/relationships`, {
            headers: { Authorization: token }
        });
        const friends = response.data;
        const friendsBar = new cliProgress.SingleBar({}, cliProgress.Presets.shades_classic);
        friendsBar.start(friends.length, 0);
        for (const friend of friends) {
            let retry = true;
            while (retry) {
                try {
                    await axios.delete(`${baseURL}/users/@me/relationships/${friend.id}`, {
                        headers: { Authorization: token }
                    });
                    log(`Removed friend: ${friend.user.username}`);
                    retry = false;
                } catch (error) {
                    if (error.response && error.response.status === 429) {
                        const retryAfter = error.response.headers['retry-after'];
                        log(`Rate limit hit. Retrying after ${retryAfter} seconds...`);
                        await new Promise((resolve) => setTimeout(resolve, retryAfter * 1000));
                    } else {
                        log(`Error removing friend: ${error.message}`);
                        retry = false;
                    }
                }
            }
            friendsBar.increment();
        }
        friendsBar.stop();
    } catch (error) {
        log(`Error removing friends: ${error.message}`);
    }
}

async function fullNuke() {
    log("Executing full nuke...");
    await processGuilds();
    await closeAllDMs();
    await removeAllFriends();
}

async function executeAction(option) {
    const confirmed = await confirmChoice();
    if (!confirmed) {
        main();
        return;
    }
    switch (option) {
        case '1':
            await displayInfo();
            break;
        case '2':
            await processGuilds();
            break;
        case '3':
            await removeAllFriends();
            break;
        case '4':
            await closeAllDMs();
            break;
        case '5':
            await fullNuke();
            break;
        default:
            console.log("Invalid option.");
    }
    console.log("\nReturning to the main menu...\n");
    main();
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function main() {
    displayBanner();
    rl.question('Choose an option (1-5): ', (option) => {
        rl.question('Enter your token: ', async (inputToken) => {
            token = inputToken;
            log(`User selected option ${option}`);
            await executeAction(option);
        });
    });
}

main();
